api/clients/accountant.go

```diff
@@ -0,0 +1,169 @@
+	//TODO: client storage; currently every instance starts fresh but on-chain or a small store makes more sense
+	// Also client is currently responsible for supplying network params, we need to add RPC in order to be automatic
+	// There's a subsequent PR that handles populating the accountant with on-chain state from the disperser
+	a := Accountant{
+		reservation:       reservation,
+		onDemand:          onDemand,
+		reservationWindow: reservationWindow,
+		pricePerSymbol:    pricePerSymbol,
+		minNumSymbols:     minNumSymbols,
+		binRecords:        []BinRecord{{Index: 0, Usage: 0}, {Index: 1, Usage: 0}, {Index: 2, Usage: 0}},
+		cumulativePayment: big.NewInt(0),
+		paymentSigner:     paymentSigner,
+	}
+	// TODO: add a routine to refresh the on-chain state occasionally?
+	return &a
+}
+
+// BlobPaymentInfo calculates and records payment information. The accountant
+// will attempt to use the active reservation first and check for quorum settings,
+// then on-demand if the reservation is not available. The returned values are
+// bin index for reservation payments and cumulative payment for on-demand payments,
+// and both fields are used to create the payment header and signature
+func (a *Accountant) BlobPaymentInfo(ctx context.Context, numSymbols uint64, quorumNumbers []uint8) (uint32, *big.Int, error) {
+	now := time.Now().Unix()
+	currentBinIndex := meterer.GetBinIndex(uint64(now), a.reservationWindow)
+
+	a.usageLock.Lock()
+	defer a.usageLock.Unlock()
+	relativeBinRecord := a.GetRelativeBinRecord(currentBinIndex)
+	relativeBinRecord.Usage += numSymbols
+
+	// first attempt to use the active reservation
+	binLimit := a.reservation.SymbolsPerSec * uint64(a.reservationWindow)
+	if relativeBinRecord.Usage <= binLimit {
+		if err := QuorumCheck(quorumNumbers, a.reservation.QuorumNumbers); err != nil {
+			return 0, big.NewInt(0), err
+		}
+		return currentBinIndex, big.NewInt(0), nil
+	}
+
+	overflowBinRecord := a.GetRelativeBinRecord(currentBinIndex + 2)
+	// Allow one overflow when the overflow bin is empty, the current usage and new length are both less than the limit
+	if overflowBinRecord.Usage == 0 && relativeBinRecord.Usage-numSymbols < binLimit && numSymbols <= binLimit {
+		overflowBinRecord.Usage += relativeBinRecord.Usage - binLimit
+		if err := QuorumCheck(quorumNumbers, a.reservation.QuorumNumbers); err != nil {
+			return 0, big.NewInt(0), err
+		}
+		return currentBinIndex, big.NewInt(0), nil
+	}
+
+	// reservation not available, attempt on-demand
+	//todo: rollback later if disperser respond with some type of rejection?
+	relativeBinRecord.Usage -= numSymbols
+	incrementRequired := big.NewInt(int64(a.PaymentCharged(uint(numSymbols))))
+	a.cumulativePayment.Add(a.cumulativePayment, incrementRequired)
+	if a.cumulativePayment.Cmp(a.onDemand.CumulativePayment) <= 0 {
+		if err := QuorumCheck(quorumNumbers, []uint8{0, 1}); err != nil {
```

Comment:
I didn't intend on reading this from disperser endpoint. Everything else made sense to, but the quorum numbers were not included in the next PR because I was expecting them to stay as 0 and 1. alternatively, I could mark them as a hard coded constant and/or add to accountant struct.

Reply:
I think either of these is fine. In the event that this set increased, it seems like it wouldn't be a breaking change for clients still using {0,1}. Maybe just a hard-coded constant for legibility would be good!